db/skipped_sequence.go

```diff
@@ -0,0 +1,211 @@
+// time in unix time
+func NewSingleSkippedSequenceEntry(sequence uint64) *SingleSkippedSequence {
+	return &SingleSkippedSequence{
+		seq:       sequence,
+		timestamp: time.Now().Unix(),
+	}
+}
+
+// getTimestamp returns the timestamp of the entry
+func (s *SingleSkippedSequence) getTimestamp() int64 {
+	return s.timestamp
+}
+
+// getLastSeq gets the last sequence in the range on the entry, for single items the sequence will be returned
+func (s *SingleSkippedSequence) getLastSeq() uint64 {
+	return s.seq
+}
+
+// setLastSeq sets the last sequence in the range on the entry, for single items its a no-op
+func (s *SingleSkippedSequence) setLastSeq(seq uint64) {
+	// no-op
+}
+
+// getStartSeq gets the start sequence in the range on the entry, for single items the sequence will be returned
+func (s *SingleSkippedSequence) getStartSeq() uint64 {
+	return s.seq
+}
+
+// isRange returns true if the entry is a sequence range entry, false if not
+func (s *SingleSkippedSequence) isRange() bool {
+	return false
+}
+
+// IsSkipped returns true if a given sequence exists in the skipped sequence slice
+func (s *SkippedSequenceSlice) IsSkipped(x uint64) bool {
+	s.lock.RLock()
+	defer s.lock.RUnlock()
+	_, found := s.findSequence(x)
+	return found
+}
+
+// SkippedSequenceCompact will compact the entries with timestamp old enough. It will also clip
+// the capacity of the slice to length + 100 if the current capacity is 2.5x the length
+func (s *SkippedSequenceSlice) SkippedSequenceCompact(ctx context.Context, maxWait int64) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	timeNow := time.Now().Unix()
+	indexToDelete := -1
+	for _, v := range s.list {
+		timeStamp := v.getTimestamp()
+		if (timeNow - timeStamp) >= maxWait {
+			indexToDelete++
+		} else {
+			// exit early to avoid iterating through whole slice
+			break
+		}
+	}
+	// if we have incremented the index counter, we need to the index range from the slice
+	if indexToDelete > -1 {
+		s.list = slices.Delete(s.list, 0, indexToDelete+1)
+	}
+	// resize slice to reclaim memory if we need to
+	s.clip(ctx)
+}
+
+// clip will clip the capacity of the slice to a constant value if the current capacity of the slice in 2.5x the length
+func (s *SkippedSequenceSlice) clip(ctx context.Context) {
+	// threshold is 2.5x the current length of the slice
+	threshold := 2.5 * float64(len(s.list))
+
+	if cap(s.list) > int(threshold) {
+		// check if we can safely clip without an out of bound errors
+		if (len(s.list) + ClipCapacityHeadroom) < cap(s.list) {
+			base.DebugfCtx(ctx, base.KeyCache, "clipping skipped list capacity")
+			s.list = s.list[:len(s.list):ClipCapacityHeadroom]
+		}
+	}
+}
+
+// findSequence will use binary search to search the elements in the slice for a given sequence
+func (s *SkippedSequenceSlice) findSequence(x uint64) (int, bool) {
+	i, found := slices.BinarySearchFunc(s.list, x, func(a SkippedSequenceListInterface, seq uint64) int {
```

Comment:
Is there any difference from a compiler perspective if the function is defined as a standalone function, as opposed to inline as you've got it?

Reply:
Have moved the definition of the search func into it's own function. All seems fine with that approach 😄